tests/testthat/test-string_boundary_linter.R

```diff
@@ -0,0 +1,136 @@
+  lintr::expect_lint(
+    "grepl('^a', x)",
+    rex::rex("Use startsWith() to detect a fixed initial substring."),
+    string_boundary_linter()
+  )
+  lintr::expect_lint(
+    "grepl('a$', x)",
+    rex::rex("Use endsWith() to detect a fixed terminal substring."),
+    string_boundary_linter()
+  )
+
+  # options besides ignore.case also don't matter
+  lintr::expect_lint(
+    "grepl('^a', x, perl = TRUE, fixed = TRUE)",
+    rex::rex("Use startsWith() to detect a fixed initial substring."),
+    string_boundary_linter()
+  )
+  # explicit FALSE (i.e., an explicit default) is ignored
+  lintr::expect_lint(
+    "grepl('^a', x, ignore.case = FALSE)",
+    rex::rex("Use startsWith() to detect a fixed initial substring."),
+    string_boundary_linter()
+  )
+})
+
+test_that("string_boundary_linter blocks simple disallowed str_detect() usages", {
+  lintr::expect_lint(
+    "str_detect(x, '^a')",
+    rex::rex("Use startsWith() to detect a fixed initial substring."),
+    string_boundary_linter()
+  )
+  lintr::expect_lint(
+    "str_detect(x, 'a$')",
+    rex::rex("Use endsWith() to detect a fixed terminal substring."),
+    string_boundary_linter()
+  )
+})
+
+test_that("string_boundary_linter blocks disallowed substr()/substring() usage", {
+  lintr::expect_lint(
+    "substr(x, 1L, 2L) == 'ab'",
+    rex::rex("Use startsWith() to detect an initial substring."),
+    string_boundary_linter()
+  )
+  # end doesn't matter, just anchoring to 1L
+  lintr::expect_lint(
+    "substr(x, 1L, end) == 'ab'",
+    rex::rex("Use startsWith() to detect an initial substring."),
+    string_boundary_linter()
+  )
+  lintr::expect_lint(
+    "substring(x, nchar(x) - 4L, nchar(x)) == 'abcde'",
+    rex::rex("Use endsWith() to detect a terminal substring."),
+    string_boundary_linter()
+  )
+  # start doesn't matter, just anchoring to nchar(x)
+  lintr::expect_lint(
+    "substring(x, start, nchar(x)) == 'abcde'",
+    rex::rex("Use endsWith() to detect a terminal substring."),
+    string_boundary_linter()
+  )
+  # more complicated expressions
+  lintr::expect_lint(
+    "substring(colnames(x), start, nchar(colnames(x))) == 'abc'",
+    rex::rex("Use endsWith() to detect a terminal substring."),
+    string_boundary_linter()
+  )
+})
+
+test_that("plain ^ or $ are skipped", {
+  lintr::expect_lint('grepl("^", x)', NULL, string_boundary_linter())
+  lintr::expect_lint('grepl("$", x)', NULL, string_boundary_linter())
+})
+
+test_that("substr inverted tests are caught as well", {
+  lintr::expect_lint(
+    "substr(x, 1L, 2L) != 'ab'",
+    rex::rex("Use startsWith() to detect an initial substring."),
+    string_boundary_linter()
+  )
+  lintr::expect_lint(
+    "substring(x, nchar(x) - 4L, nchar(x)) != 'abcde'",
+    rex::rex("Use endsWith() to detect a terminal substring."),
+    string_boundary_linter()
+  )
+})
+
+# TODO(chiricom): extend to inversion cases too:
+#  !grepl("^abc", x) --> !startsWith(x, "abc")
```

Comment:
What would need to change?

Reply:
Indeed it's trivially ours already; added a test. Maybe I was thinking of the `substr() != ...` case 🤔

Reply:
(but that's covered already too)